WT-6914 Write database files subpage for architecture guide

mm-ng opened the pull request.

ddanderson reviewed (commented):
@mm-ng, this is a great start.  I think you've covered this in about the right depth.  I've made a few specific suggestions.  I was thinking that it would be good to mention a few of the important functions used in the creation/reading of data files:

Low level functions that pack or unpack the contents of individual cells are in `cell_inline.h` .  Functions like `__wt_cell_unpack_kv` and `__wt_cell_unpack_addr` to unpack keys values and addresses are found here, and uses of such functions appear in the BTree code.  Similarly, there are `__wt_cell_pack_*` functions to pack particular kinds of keys, values and addresses; these are used in the reconciliation code.

You can use my wording (or not).  This change basically LGTM, but I'll take another review pass when you're ready.

ddanderson commented on src/docs/arch-data-file.dox:
```diff
@@ -1,17 +1,90 @@
+internal_key_max, leaf_key_max and leaf_value_max are used to determine 
+overflow items.
+
+Whilst the above gives a general overview of how different page types are 
```
Whilst -> while  (whilst is archaic in US English).  I think this appears in a few spots.

mm-ng replied:
Thanks for the heads up, I didn't know this!

mm-ng commented on src/docs/arch-data-file.dox:
```diff
@@ -1,17 +1,90 @@
+Database Files
+
+WiredTiger stores databases in a durable format through writing data to the 
+physical disk/s. As part of this process, WiredTiger creates a number of 
```
Yes that makes sense, I will change this.

mm-ng commented on src/docs/arch-data-file.dox:
```diff
@@ -1,17 +1,90 @@
+data being stored, WiredTiger may compress data to preserve memory and disk 
+space. Keys may not be stored in their entirety when prefix compression 
+is used (the identical key prefix is stored only once per page to reduce 
+size).
```
Good point.

mm-ng commented:
@ddanderson Thanks so much for all the feedback. I have incorporated your suggestions and also added a section on the data file functions. Let me know if there is anything else you think would be worth discussing in the article.

ddanderson commented on src/docs/arch-data-file.dox:
```diff
@@ -1,17 +1,105 @@
+- The row-store internal page \c WT_PAGE_ROW_INT is made up of a key cell
+(\c WT_CELL_KEY or \c WT_CELL_KEY_OVFL) followed by an offpage-reference
+(\c WT_CELL_ADDR_XXX).
+- The column-store internal page simply contains an off-page reference.
```
Is this true?  References don't have a corresponding starting (recordid) key?

mm-ng replied:
What I meant by this sentence is I thought the column-store internal page only had a WT_CELL_ADDR_<> cell. I believe the recordid is part of the WT_CELL_COMMON_FIELDS which is shared across all cell types. Would it be helpful to talk about the structure of cells?

ddanderson replied:
You're right, it looks like internal pages just have references (presumably the key needed for searching can be deduced by looking at the first key on each page referenced).

ddanderson commented on src/docs/arch-data-file.dox:
```diff
@@ -1,17 +1,105 @@
+- The row-store leaf page \c WT_PAGE_ROW_LEAF is made up of a key cell
+(\c WT_CELL_KEY or \c WT_CELL_KEY_OVFL) followed by a value
+(\c WT_CELL_VALUE/VALUE_COPY/VALUE_OVFL).
+- The column-store leaf page (for variable-length data) WT_PAGE_COL_VAR
+is made up of data cells (WT_CELL_VALUE/VALUE_COPY/VALUE_OVFL) or
+deleted cells (WT_CELL_DEL)
```
If there is no record_id key, how is it deduced?  (There must be a starting id somewhere).

ddanderson replied:
You might add: `The page header has a starting record number, so the associated keys for all column store values is deduced by their position in the page.`

mm-ng replied:
Thanks for this suggestion Don. I have added it in and will make the merge.

ddanderson reviewed (approved):
LGTM.  I have some remaining questions about how column keys are deduced.

ddanderson commented:
@mm-ng, I've asked you to insert one sentence.  Everything else looks great, ready to merge.
